crates/cargo-test-support/src/registry.rs

```diff
@@ -611,16 +638,72 @@ impl HttpServer {
                     self.dl(&req)
                 }
             }
+            // publish
+            ("put", ["api", "v1", "crates", "new"]) => {
+                if !authorized(true) {
+                    return self.unauthorized(req)
+                }
+                if let Some(body) = &req.body {
+                    // Get the metadata of the package
+                    let (len, remaining)  = body.split_at(4);
+                    let json_len = u32::from_le_bytes(len.try_into().unwrap());
+                    let (json, remaining) = remaining.split_at(json_len as usize);
+                    let new_crate = serde_json::from_slice::<crates_io::NewCrate>(json).unwrap();
+                    // Get the `.crate` file
+                    let (len, remaining) = remaining.split_at(4);
+                    let file_len = u32::from_le_bytes(len.try_into().unwrap());
+                    let (file, _remaining) = remaining.split_at(file_len as usize);
+
+                    // Write the `.crate`
+                    let dst = self.dl_path.join(&new_crate.name).join(&new_crate.vers).join("download");
+                    t!(fs::create_dir_all(dst.parent().unwrap()));
+                    let mut f = t!(File::create(&dst));
+                    t!(f.write_all(file));
+
+                    let deps = new_crate.deps.iter().map(|dep| {
+                        serde_json::json!({
+                            "name": dep.name,
```

Comment:
Yea, I believe this and `package` aren't quite correct.

The `name` and `package` fields change depending on whether or not `explicit_name_in_toml` is set. The general logic (stolen from crates.io) looks like:

```rust
let (name, package) = match &dep.explicit_name_in_toml {
    Some(explicit) => (explicit.to_string(), Some(dep.name.to_string())),
    None => (dep.name.to_string(), None),
};
```